File: algorithms/hard/maximum_score_words_formed_by_letters.ex
defmodule Solution do
  @spec max_score_words(words :: [String.t()], letters :: [char], score :: [integer]) :: integer
  def max_score_words(words, letters, score) do
    words = Enum.map(words, &String.to_charlist/1)
    dfs(words, to_map(letters), score_map(score), 0, 0)
  end

  def dfs([], _letters, _score, curr, ans), do: max(curr, ans)

  def dfs([w | ws], letters, score, curr, ans) do
    if can_build_word(w, letters) do
      s = calc_score(w, score)
      letters = update_letters(w, letters, &(&1 - 1))
      ans = max(dfs(ws, letters, score, curr + s, ans), ans)
      letters = update_letters(w, letters, &(&1 + 1))
      dfs(ws, letters, score, curr, ans)
    else
      dfs(ws, letters, score, curr, ans)
    end
  end

  def update_letters(w, letters, f) do
    Enum.reduce(w, letters, fn c, m ->
      Map.update!(m, c, f)
    end)
  end

  def calc_score(w, score) do
    Enum.reduce(w, 0, fn c, acc ->
      acc + score[c]
    end)
  end

  def can_build_word(w, letters) do
    w
    |> to_map()
    |> Enum.all?(fn {k, v} ->
      Map.get(letters, k, 0) >= v
    end)
  end

  def to_map(lst) do
    lst
    |> Enum.reduce(%{}, fn c, m ->
      Map.update(m, c, 1, &(&1 + 1))
    end)
  end

  def score_map(score) do
    Enum.zip(?a..?z, score)
    |> Enum.reduce(%{}, fn {c, v}, m ->
      Map.put(m, c, v)
    end)
  end
end

File: algorithms/hard/maximum_score_words_formed_by_letters.rs
use std::collections::HashMap;

fn can_build_word(s: &str, lettersMap: &HashMap<char, i32>) -> bool {
    let mut ws = HashMap::new();
    for c in s.chars() {
        *ws.entry(c).or_insert(0) += 1;
    }

    for (k, v) in &ws {
        if !lettersMap.contains_key(k) {
            return false;
        }
        if lettersMap[k] < *v {
            return false;
        }
    }
    true
}

fn calc_score(s: &str, score: &[i32]) -> i32 {
    s.chars().fold(0, |acc, c| acc + score[(c as u8 - b'a') as usize])
}

fn update_letters(s: &str, lettersMap: &mut HashMap<char, i32>, added: bool) {
    s.chars().for_each(|c| {
        lettersMap.entry(c).and_modify(|v| {
            if added {
                *v += 1;
            } else {
                *v -= 1;
            }
        });
    });
}

fn dfs(words: &[String], lettersMap: &mut HashMap<char, i32>, score: &[i32], curr: i32, ans: &mut i32) {
    if words.is_empty() {
        if *ans < curr {
            *ans = curr;
        }
        return;
    }

    if can_build_word(&words[0], lettersMap) {
        let s = calc_score(&words[0], score);
        update_letters(&words[0], lettersMap, false);
        dfs(&words[1..], lettersMap, score, curr + s, ans);
        update_letters(&words[0], lettersMap, true);
    }
    
    dfs(&words[1..], lettersMap, score, curr, ans);
}

impl Solution {
    pub fn max_score_words(words: Vec<String>, letters: Vec<char>, score: Vec<i32>) -> i32 {
        let mut lettersMap = HashMap::new();
        for &c in &letters {
            *lettersMap.entry(c).or_insert(0) += 1;
        }
        let mut ans = 0;
        dfs(&words, &mut lettersMap, &score, 0, &mut ans);
        ans
    }
}
